### PR TITLE
drivers: Add flash driver for RRAM

### PR DESCRIPTION
**Pull request by jaz1-nordic**
Added a simple driver for RRAM. It is implemented as a flash driver,
because the "RRAM eFlash" macro obeys flash-like constraints.
Although users are not required to erase before write.

### CODE REVIEW

File: drivers/flash/soc_flash_nrf_rram.c [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+#include <zephyr/../../drivers/flash/soc_flash_nrf.h>
+
+LOG_MODULE_REGISTER(flash_nrf_rram, CONFIG_FLASH_LOG_LEVEL);
+
+#define RRAM DT_INST(0, soc_nv_flash)
+
+#define RRAM_START DT_REG_ADDR(RRAM)
+#define RRAM_SIZE  DT_REG_SIZE(RRAM)
+
+#define RRAM_WORD_SIZE 4
```
[the remark this answers is not in the document]
Comment (de-nordic) [reply drivers/flash/soc_flash_nrf_rram.c:27]: Do we allow partial writes?

File: drivers/flash/soc_flash_nrf_rram.c [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+#define WRITE_LINE_SIZE     RRAM_WORD_SIZE
+#endif
+
+#ifndef CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE
+#define FLASH_SLOT_WRITE 5000
+
+#if WRITE_BUFFER_ENABLE
+#define RRAM_MAX_WRITE_BUFFER (WRITE_BUFFER_SIZE * WRITE_LINE_SIZE)
+#else
+#define RRAM_MAX_WRITE_BUFFER (WRITE_LINE_SIZE * 32)
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf_rram.c:55]: Why?

File: drivers/flash/soc_flash_nrf_rram.c [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+#if WRITE_BUFFER_ENABLE
+static void commit_changes(size_t len)
+{
+	if (nrf_rramc_empty_buffer_check(NRF_RRAMC)) {
+		/* The internal write-buffer has been committed to RRAM and is now empty. */
+		return;
+	}
+
+	if ((len % (WRITE_LINE_SIZE * WRITE_BUFFER_SIZE)) == 0) {
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf_rram.c:78]: Is that RRAM_MAX_WRITE_BUFFER ?

File: drivers/flash/soc_flash_nrf_rram.c [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+static int flash_nrf_write(off_t addr, const void *data, size_t len)
+{
+	int ret = 0;
+
+	if (!is_within_bounds(addr, len, 0, RRAM_SIZE)) {
+		return -EINVAL;
+	}
+	addr += RRAM_START;
+
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf_rram.c:181]: There is not write alignment check. Do we allow to write `len` that is not write_block_size aligned?
Flash API does not allow that.

File: drivers/flash/soc_flash_nrf_rram.c [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+	size_t len;
+
+	uint32_t i = 0U;
+
+	if (w_ctx->enable_time_limit) {
+		nrf_flash_sync_get_timestamp_begin();
+	}
+
+	while (w_ctx->len > 0) {
+		len = (RRAM_MAX_WRITE_BUFFER < w_ctx->len) ? RRAM_MAX_WRITE_BUFFER : w_ctx->len;
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf_rram.c:136]: If we are in agreement with Flash API, so we write by `write_block_size`, then we either write by full buffers or `write_block_size`

File: drivers/flash/soc_flash_nrf_rram.c [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+	barrier_dmem_fence_full();
+}
+#endif
+
+static void rram_write(off_t addr, const void *data, size_t len)
+{
+	nrf_rramc_config_t config = {.mode_write = true, .write_buff_size = WRITE_BUFFER_SIZE};
+
+	nrf_rramc_config_set(NRF_RRAMC, &config);
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf_rram.c:93]: Shouldn't the init do something like that to set the default 'read' mode?

**Review by de-nordic — CHANGES_REQUESTED** [review]
Can we drop `flash_` prefix, in case of static functions, wherever it is possible?

File: drivers/flash/soc_flash_nrf_rram.c (hunk 6 above)
Comment (jaz1-nordic) [reply drivers/flash/soc_flash_nrf_rram.c:93]: "Read only" is the standard mode after reset. Additionally, after each write, the mode is restored to "read only".

File: drivers/flash/soc_flash_nrf_rram.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaz1-nordic) [reply drivers/flash/soc_flash_nrf_rram.c:27]: If the write buffer is disabled, the written data (byte, half-word, 32-bit, or 64-bit) is committed to RRAM immediately. So, yes.

File: drivers/flash/soc_flash_nrf_rram.c (hunk 4 above)
Comment (jaz1-nordic) [reply drivers/flash/soc_flash_nrf_rram.c:181]: write_block_size is set to 1.

File: drivers/flash/soc_flash_nrf_rram.c (hunk 5 above)
Comment (jaz1-nordic) [reply drivers/flash/soc_flash_nrf_rram.c:136]: write_block_size is set to 1.

File: drivers/flash/soc_flash_nrf_rram.c (hunk 2 above)
Comment (jaz1-nordic) [reply drivers/flash/soc_flash_nrf_rram.c:55]: @nvlsianpu Could you explain? This has a strange value to me as well.

File: drivers/flash/soc_flash_nrf_rram.c [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+	struct flash_op_desc flash_op_desc = {.handler = write_op, .context = &context};
+
+	nrf_flash_sync_set_context(FLASH_SLOT_WRITE);
+	return nrf_flash_sync_exe(&flash_op_desc);
+}
+
+#endif /* !CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE */
+
+static int nrf_write(off_t addr, const void *data, size_t len)
+{
```
Comment (Laczen) [inline drivers/flash/soc_flash_nrf_rram.c:183]: check for `data != NULL` and return -EINVAL, this keeps the meaning of `data == NULL` for delete internal to the driver.

File: drivers/flash/soc_flash_nrf_rram.c (hunk 2 above)
Comment (jaz1-nordic) [reply drivers/flash/soc_flash_nrf_rram.c]: According to information from @nvlsianpu, he meant the size of the native RRAM buffer (128-bits, not bytes). Changed.

File: drivers/flash/soc_flash_nrf_rram.c (hunk 4 above)
Comment (de-nordic) [reply drivers/flash/soc_flash_nrf_rram.c:196]: But RRAM, that we have has wbs=16

**Review by de-nordic — APPROVED** [review]

**Review by Laczen — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**